etl/airflow.md

```diff
@@ -0,0 +1,329 @@
+
+    - `start_date`: the date that the DAG should begin running tasks for. For
+    example, if you intend to run a task to run a scraping task daily starting
+    on January 1, 2020, the first DAG run will happen on January 2 in order to
+    gather all of the information that may have been added to the source
+    throughout the day of January 1.
+    - `end_date`: the last date for which a DAG should run.
+    - `execution_timeout`: How long the task should continue to run before
+    giving up. For example, if you need a DAG to run every 3 hours, you may want
+    to make sure the previous DAG run is no longer running. Setting an
+    `execution_timeout` of, for example, 2 hours and 55 minutes, would force the
+    first DAG run to stop in time for the next DAG run to begin without
+    overlapping.
+    - `retries`: If a DAG fails, the number of times it should try to run again.
+
+3. **Instantiate a DAG:** The next step is to instantiate a DAG. Here are a
+couple of important arguments to know:
+
+    - `dag_id`: a string that gives the DAG a name. This is the name that shows
+    up in the Airflow dashboard for this DAG.
+    - `schedule_interval`: the amount of time that should elapse between DAG
+    runs. This can be [a cron string](https://crontab.guru/), or you can use
+    [one of Airflow’s presets](http://airflow.apache.org/docs/stable/dag-run.html#cron-presets)
+    - `default_args`: set this equal to the dictionary defined in step 2. More
+    info about [the Airflow DAG object](http://airflow.apache.org/docs/stable/_api/airflow/models/dag/index.html#module-airflow.models.dag)
+
+4. **Tasks:** In the penultimate step of defining your pipeline, define out the
+DAG’s tasks. Each task is an instance of one of Airflow’s operators, so look at
+the documentation for that specific operator for the most detailed information
+on how to define a task. Here’s the list of operators:
+[https://airflow.apache.org/docs/stable/_api/airflow/operators/index.html](https://airflow.apache.org/docs/stable/_api/airflow/operators/index.html)
+
+    Keep in mind that there are also operators that are focused on control flow
+    that you might need to use in defining tasks. A couple common ones you might
+    use:
+
+    - Branch Operator: Allows you to run different tasks at different times or
+    under different conditions
+    - Dummy Operator: If you want to avoid running a task if a certain condition
+    is true, you can use the Dummy Operator to run a fake task in place of the
+    true one.
+
+5. **Task ordering and dependencies:** The final part of the pipeline is task
+ordering. Here, set the order that the tasks should run in and how they depend
+on each other. The example code given on the Airflow docs is concise and
+helpful:
+[https://airflow.apache.org/docs/stable/tutorial.html#setting-up-dependencies](https://airflow.apache.org/docs/stable/tutorial.html#setting-up-dependencies)
+
+## Recommended default settings
```

Comment:
@jeancochrane Fatima indicated this section might be incomplete. Can you review?

Reply:
Good catch on the connection string @fatima3558!

I just ran some diffs between the [default config for Airflow 1.10.10](https://raw.githubusercontent.com/apache/airflow/96697180d79bfc90f6964a8e99f9dd441789177c/airflow/config_templates/default_airflow.cfg) and our local config (`airflow.cfg`) and production config (`configs/airflow.production.cfg`) for https://github.com/datamade/la-metro-dashboard/. Here are some other changes that I found:

### Local `airflow.cfg`

```cfg
# Don't load example DAGs
load_examples = False

# Turn on authentication
authenticate = True

# Turn on role-based access control (Airflow's new model of authentication, with a separate UI)
rbac = True

# Don't catch up to old DAGs when restarting the process
catchup_by_default = False
```

We may not always want authentication and RBAC. I think that should probably be a project-by-project decision.

We also changed `executor` to `LocalExecutor` from `SequentialExecutor` but I don't actually think we should do this going forward because [`SequentialExecutor` is just `LocalExecutor` with parallelism set to `1`, which is probably good for local testing](https://airflow.apache.org/docs/stable/executor/local.html).

### Production `configs/airflow.production.cfg`

```cfg
# Store logs in /var/log intead of AIRFLOW_HOME
base_log_folder = /var/log/la-metro-dashboard/airflow
dag_processor_manager_log_location = /var/log/la-metro-dashboard/airflow/dag_processor_manager/dag_processor_manager.log

# Use a single node for task execution
executor = LocalExecutor

# Only two task instances should be able to run simultaneously
parallelism = 2

# The scheduler should only be able to run two task instances
dag_concurrency = 2

# Workers should only be able to run two task instances
worker_concurrency = 2

# Don't load example DAGs
load_examples = False

# Turn on authentication
authenticate = True

# Turn on role-based access control (Airflow's new model of authentication, with a separate UI)
rbac = True

# Don't catch up to old DAGs when restarting the process
catchup_by_default = False
```

Again, authentication and RBAC should be set project-by-project. The concurrency configs are super confusing and [the docs](https://airflow.apache.org/docs/stable/configurations-ref.html) don't really help so I think there's a lot of opportunity for better tuning in the future.